src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -500,3 +503,102 @@ export async function refundRequestIsValid(
+};
+
+// @description Search for fills recorded by a specific SpokePool.
+// @param chainId Chain ID of the relevant SpokePoolClient instance.
+// @param spokePoolClients Set of SpokePoolClient instances, mapped by chain ID.
+// @param filter  Optional filtering criteria.
+// @returns Array of FillWithBlock events matching the chain ID and optional filtering criteria.
+export async function getFills(
+  chainId: number,
+  spokePoolClients: SpokePoolClients,
+  filter: SpokePoolFillFilter = {}
+): Promise<FillWithBlock[]> {
+  const spokePoolClient = spokePoolClients[chainId];
+  assert(isDefined(spokePoolClient));
+
+  const { originChainId, repaymentChainId, relayer, isSlowRelay, fromBlock } = filter;
+
+  const fills = await filterAsync(spokePoolClient.getFills(), async (fill) => {
+    if (isDefined(fromBlock) && fill.blockNumber > fromBlock) {
+      return false;
+    }
+
+    // @dev tsdx and old Typescript seem to prevent dynamic iteration over the filter, so evaluate the keys manually.
+    if (
+      (isDefined(originChainId) && fill.originChainId !== originChainId) ||
+      (isDefined(repaymentChainId) && fill.repaymentChainId !== repaymentChainId) ||
+      (isDefined(relayer) && fill.relayer !== relayer) ||
+      (isDefined(isSlowRelay) && fill.updatableRelayData.isSlowRelay !== isSlowRelay)
+    ) {
+      return false;
+    }
+
+    // @dev The SDK-v2 UBAClient stores the base SpokePoolClient definition, but here we use an extended variant.
+    // This will be resolved when upstreaming to SDK-v2.
+    const originSpokePoolClient = spokePoolClients[fill.originChainId];
+    if (!isDefined(originSpokePoolClient)) {
+      return false;
+    }
+
+    const deposit = await queryHistoricalDepositForFill(originSpokePoolClient, fill);
+    return isDefined(deposit);
+  });
+
+  return fills;
+}
+
+// @description Search for refund requests recorded by a specific SpokePool.
+// @param chainId Chain ID of the relevant SpokePoolClient instance.
+// @param chainIdIndices Complete set of ordered chain IDs.
+// @param hubPoolClient HubPoolClient instance.
+// @param spokePoolClients Set of SpokePoolClient instances, mapped by chain ID.
+// @param filter  Optional filtering criteria.
+// @returns Array of RefundRequestWithBlock events matching the chain ID and optional filtering criteria.
+export async function getRefundRequests(
+  chainId: number,
+  chainIdIndices: number[],
+  hubPoolClient: HubPoolClient,
+  spokePoolClients: SpokePoolClients,
+  filter: SpokePoolEventFilter = {}
+): Promise<RefundRequestWithBlock[]> {
+  const spokePoolClient = spokePoolClients[chainId];
+  assert(isDefined(spokePoolClient));
+
+  const { originChainId, destinationChainId, relayer, fromBlock } = filter;
+
+  const refundRequests = await filterAsync(spokePoolClient.getRefundRequests(), async (refundRequest) => {
+    assert(refundRequest.repaymentChainId === chainId);
+
+    if (isDefined(fromBlock) && refundRequest.blockNumber > fromBlock) {
```

Comment:
For this check and the similar check in `getFills`, should we be checking that `refund/fill.blockNumber >= spokePoolClient.deploymentBlockNumber`?

Reply:
@nicholaspai Is this perhaps a check we should apply down in the SpokePoolClient `update()` method instead? It seems safer to just block such invalid data from ever being stored, and it has the benefit of only needing to be evaluated once, versus every time we request the array of fills.

Reply:
yea that makes sense